File: packages/sdk-typescript/tests/unit/egress.test.ts
// SDK unit tests for EgressResource.
//
// Covers the contract surface:
// - attachToSession sends POST /v1/sessions/{id}/proxy with the
//   SessionEgressConfig body verbatim (URL-encoded id).
// - getSessionProxy reads GET on the same path.
// - the saved-proxy library CRUD + test rides the LIVE account-proxies
//   API: createProxy/listProxies POST/GET /v1/account/me/proxies,
//   updateProxy PUT + deleteProxy DELETE /v1/account/me/proxies/{id},
//   testProxy POST /v1/account/me/proxies/{id}/test.

import { describe, expect, it } from 'vitest';
import { EgressResource } from '../../src/resources/egress.js';
import type { HttpClient } from '../../src/http.js';
import type { SessionEgressConfig, AccountProxyInput } from '@driftstack/api-types';

interface RecordedRequest {
  method: string;
  path: string;
  body?: unknown;
}

function makeFakeHttp<T>(reply: T): { http: HttpClient; calls: RecordedRequest[] } {
  const calls: RecordedRequest[] = [];
  const http = {
    request: <R>(opts: { method: string; path: string; body?: unknown }) => {
      const recorded: RecordedRequest = { method: opts.method, path: opts.path };
      if (opts.body !== undefined) recorded.body = opts.body;
      calls.push(recorded);
      return Promise.resolve(reply as unknown as R);
    },
  } as unknown as HttpClient;
  return { http, calls };
}

describe('EgressResource', () => {
  it('attachToSession POSTs /v1/sessions/{id}/proxy with the body verbatim (id URL-encoded)', async () => {
    const config: SessionEgressConfig = {
      session_id: 'ses xyz',
      proxy: {
        type: 'socks5',
        socks5: {
          host: 'proxy.example.com',
          port: 1080,
          udp_associate: true,
          require_remote_dns: false,
        },
      },
      egress_safeguard: {
        block_direct_internet: true,
        block_unproxied_dns: true,
        block_webrtc_stun_leakage: true,
      },
    };
    const reply = {
      type: 'socks5' as const,
      safeguards: {
        block_direct_internet: true,
        block_unproxied_dns: true,
        block_webrtc_stun_leakage: true,
      },
    };
    const { http, calls } = makeFakeHttp(reply);
    const res = new EgressResource(http);
    const result = await res.attachToSession('ses xyz', config);
    expect(calls).toEqual([{ method: 'POST', path: '/v1/sessions/ses%20xyz/proxy', body: config }]);
    expect(result).toEqual(reply);
  });

  it('getSessionProxy GETs /v1/sessions/{id}/proxy', async () => {
    const { http, calls } = makeFakeHttp({
      type: 'socks5',
      safeguards: {
        block_direct_internet: true,
        block_unproxied_dns: true,
        block_webrtc_stun_leakage: true,
      },
    });
    const res = new EgressResource(http);
    await res.getSessionProxy('ses_abc');
    expect(calls).toEqual([{ method: 'GET', path: '/v1/sessions/ses_abc/proxy' }]);
  });

  it('createProxy POSTs /v1/account/me/proxies with the flat AccountProxyInput body', async () => {
    const body: AccountProxyInput = {
      label: 'team SOCKS5',
      scheme: 'socks5',
      host: 'x.example',
      port: 1080,
      username: null,
      password: 'secret',
    };
    const reply = {
      id: 'apx_1',
      label: 'team SOCKS5',
      scheme: 'socks5' as const,
      host: 'x.example',
      port: 1080,
      username: null,
      has_password: true,
      has_secret: false,
      created_at: '2026-06-17T00:00:00Z',
      updated_at: '2026-06-17T00:00:00Z',
    };
    const { http, calls } = makeFakeHttp(reply);
    const res = new EgressResource(http);
    const result = await res.createProxy(body);
    expect(calls).toEqual([{ method: 'POST', path: '/v1/account/me/proxies', body }]);
    expect(result).toEqual(reply);
  });

  it('listProxies GETs /v1/account/me/proxies', async () => {
    const reply = { data: [{ id: 'apx_1', label: 'l', scheme: 'socks5' as const }] };
    const { http, calls } = makeFakeHttp(reply);
    const res = new EgressResource(http);
    const result = await res.listProxies();
    expect(calls).toEqual([{ method: 'GET', path: '/v1/account/me/proxies' }]);
    expect(result.data).toHaveLength(1);
  });

  it('updateProxy PUTs /v1/account/me/proxies/{id} (URL-encoded)', async () => {
    const { http, calls } = makeFakeHttp({ id: 'p 1' });
    const res = new EgressResource(http);
    await res.updateProxy('p 1', { label: 'renamed' });
    expect(calls).toEqual([
      { method: 'PUT', path: '/v1/account/me/proxies/p%201', body: { label: 'renamed' } },
    ]);
  });

  it('deleteProxy DELETEs /v1/account/me/proxies/{id} (URL-encoded)', async () => {
    const { http, calls } = makeFakeHttp(undefined as unknown as void);
    const res = new EgressResource(http);
    await res.deleteProxy('proxy with space');
    expect(calls).toEqual([
      { method: 'DELETE', path: '/v1/account/me/proxies/proxy%20with%20space' },
    ]);
  });

  it('testProxy POSTs /v1/account/me/proxies/{id}/test', async () => {
    const { http, calls } = makeFakeHttp({ ok: true, latency_ms: 42 });
    const res = new EgressResource(http);
    const result = await res.testProxy('apx_1');
    expect(calls).toEqual([{ method: 'POST', path: '/v1/account/me/proxies/apx_1/test' }]);
    expect(result).toEqual({ ok: true, latency_ms: 42 });
  });
});

File: packages/sdk-go/egress_test.go
package driftstack

import (
	"context"
	"encoding/json"
	"net/http"
	"testing"
)

func TestEgress_AttachToSession(t *testing.T) {
	t.Parallel()
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/sessions/ses_xyz/proxy" || r.Method != "POST" {
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		var body map[string]any
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatal(err)
		}
		if body["session_id"] != "ses_xyz" {
			t.Errorf("session_id=%v", body["session_id"])
		}
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(SessionProxyAttachResponse{
			Type: "socks5",
			Safeguards: map[string]bool{
				"block_direct_internet":     true,
				"block_unproxied_dns":       true,
				"block_webrtc_stun_leakage": true,
			},
		})
	})
	got, err := client.Egress.AttachToSession(context.Background(), "ses_xyz", &SessionEgressConfig{
		SessionID: "ses_xyz",
		Proxy: map[string]any{
			"type":   "socks5",
			"socks5": map[string]any{"host": "p.example", "port": 1080, "udp_associate": true},
		},
		EgressSafeguard: map[string]bool{
			"block_direct_internet":     true,
			"block_unproxied_dns":       true,
			"block_webrtc_stun_leakage": true,
		},
	})
	if err != nil {
		t.Fatal(err)
	}
	if got.Type != "socks5" {
		t.Errorf("type=%q", got.Type)
	}
}

func TestEgress_AttachToSession_URLEscapesSessionID(t *testing.T) {
	t.Parallel()
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		// EscapedPath returns the URL-encoded form (preserves %20);
		// r.URL.Path is the decoded form.
		if r.URL.EscapedPath() != "/v1/sessions/ses%20with%20space/proxy" {
			t.Errorf("escaped path=%q", r.URL.EscapedPath())
		}
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(SessionProxyAttachResponse{Type: "socks5"})
	})
	_, err := client.Egress.AttachToSession(context.Background(), "ses with space", &SessionEgressConfig{
		SessionID: "ses with space",
		Proxy:     map[string]any{"type": "socks5"},
	})
	if err != nil {
		t.Fatal(err)
	}
}

func TestEgress_GetSessionProxy(t *testing.T) {
	t.Parallel()
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/sessions/ses_xyz/proxy" || r.Method != "GET" {
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(SessionProxyAttachResponse{Type: "socks5"})
	})
	got, err := client.Egress.GetSessionProxy(context.Background(), "ses_xyz")
	if err != nil {
		t.Fatal(err)
	}
	if got.Type != "socks5" {
		t.Errorf("type=%q", got.Type)
	}
}

func TestEgress_CreateProxy(t *testing.T) {
	t.Parallel()
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/account/me/proxies" || r.Method != "POST" {
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(AccountProxyMetadata{ID: "apx_1", Label: "team", Scheme: "socks5"})
	})
	got, err := client.Egress.CreateProxy(context.Background(), &AccountProxyInput{
		Label:  "team",
		Scheme: "socks5",
		Host:   "x.example",
		Port:   1080,
	})
	if err != nil {
		t.Fatal(err)
	}
	if got.ID != "apx_1" || got.Label != "team" {
		t.Errorf("metadata=%+v", got)
	}
}

func TestEgress_ListProxies_EmptyList(t *testing.T) {
	t.Parallel()
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/account/me/proxies" || r.Method != "GET" {
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(AccountProxyList{Data: []AccountProxyMetadata{}})
	})
	got, err := client.Egress.ListProxies(context.Background())
	if err != nil {
		t.Fatal(err)
	}
	if len(got.Data) != 0 {
		t.Errorf("expected empty list, got %d", len(got.Data))
	}
}

func TestEgress_UpdateProxy(t *testing.T) {
	t.Parallel()
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/account/me/proxies/apx_1" || r.Method != "PUT" {
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(AccountProxyMetadata{ID: "apx_1", Label: "renamed"})
	})
	got, err := client.Egress.UpdateProxy(context.Background(), "apx_1", map[string]any{"label": "renamed"})
	if err != nil {
		t.Fatal(err)
	}
	if got.Label != "renamed" {
		t.Errorf("metadata=%+v", got)
	}
}

func TestEgress_DeleteProxy(t *testing.T) {
	t.Parallel()
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/account/me/proxies/apx_1" || r.Method != "DELETE" {
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		w.WriteHeader(http.StatusNoContent)
	})
	if err := client.Egress.DeleteProxy(context.Background(), "apx_1"); err != nil {
		t.Fatal(err)
	}
}

func TestEgress_TestProxy(t *testing.T) {
	t.Parallel()
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/account/me/proxies/apx_1/test" || r.Method != "POST" {
			t.Errorf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(AccountProxyTestResult{Ok: true, LatencyMs: 42})
	})
	got, err := client.Egress.TestProxy(context.Background(), "apx_1")
	if err != nil {
		t.Fatal(err)
	}
	if !got.Ok || got.LatencyMs != 42 {
		t.Errorf("result=%+v", got)
	}
}
